src/rust/engine/fs/src/lib.rs

```diff
@@ -545,6 +545,24 @@ impl PosixFS {
   }
 }
 
+impl VFS<io::Error> for Arc<PosixFS> {
```

Comment:
Why the impl against a specific reference type and not `PosixFS` itself? Is there a technical compiler difficulty this addresses?

Reply:
So... now that I think about it, it's guaranteed not to be necessary in this case. I did this thinking that it would be relevant to consumers, but it could not be, since they could always derefence.

It _can_ be useful to implement a trait for a reference type because you know that `self` will need to be borrowed in strange ways, or recursively. But it does not here, afaik.

Reply:
...Whoops. That's a lie. VFS puts a `clone` bound on the struct it's implemented for, and cloning PosixFS would mean cloning its root... no bueno.
```
error[E0277]: the trait bound `PosixFS: std::clone::Clone` is not satisfied
   --> src/rust/engine/fs/src/lib.rs:548:6
    |
548 | impl VFS<io::Error> for PosixFS {
    |      ^^^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `PosixFS`
    |
    = note: required by `VFS`

error: aborting due to previous error
```

So yes. Necessary.

Reply:
I won't block, but this just begs the question why `VFS: Clone`. Looking at the 3 spots it's used in the default methods, its completely unclear (to me) why a clone on self is needed.

Reply:
VFS: requires `self: Clone` because it needs to execute chained filesystem calls in `Futures`, so it needs to clone the "context" (VFS impl) into the closures for the chained calls. See https://github.com/pantsbuild/pants/blob/504092d47b94d861fa95d9e7428b202fac6e9628/src/rust/engine/fs/src/lib.rs#L567